drv/lpc55-swd/build.rs

```diff
@@ -88,6 +97,163 @@ fn generate_swd_functions(config: &TaskConfig) -> Result<()> {
+
+    Ok(())
+}
+
+/// Given a reader on an ELF file return the executable image.
+pub fn get_elf<R>(mut reader: R) -> Result<Vec<u8>, anyhow::Error>
+where
+    R: Read + Seek,
+{
+    let mut data = vec![];
+    let _len = reader
+        .read_to_end(&mut data)
+        .context("cannot read to end")?;
+    let elf = Elf::parse(&data).context("cannot parse data as ELF")?;
+
+    if elf.header.container()? != Container::Little
+        || elf.header.e_machine != goblin::elf::header::EM_ARM
+    {
+        bail!("not a little-endian ARM ELF file");
+    }
+
+    // Extract the bytes to load
+
+    // Find all of the sections that compose our '.bin' file.
+    // These criteria were inferred from the produced artifact
+    // and the result of `arm-none-eabi-objcopy -O binary $INPUT_ELF $OUTPUT_BIN`
+    //
+    // See also:
+    // https://refspecs.linuxfoundation.org/LSB_2.1.0/LSB-Embedded/LSB-Embedded/elftypes.html
+    let sections: Vec<&SectionHeader> = elf
+        .section_headers
+        .iter()
+        .filter(|sh| (sh.sh_type & SHT_PROGBITS == SHT_PROGBITS))
+        .filter(|sh| (sh.sh_flags as u32 & SHF_ALLOC == SHF_ALLOC))
+        .filter(|sh| !sh.vm_range().is_empty())
+        .collect();
+    let bin_size: usize = sections.iter().map(|sh| sh.vm_range().len()).sum();
+
+    // Do a sanity check on the size of the blob based on what we've seen.
+    //
+    // Notes:
+    //
+    // Clippy ignores the profile (or equivalent) when building the
+    // endoscope blob which results in an enormous binary. Since we're not
+    // going to actually build the swd task when running clippy, the large
+    // size can be ignored.
+    //
+    // The artifact-specific profiles do not allow `lto` to be
+    // specified. That would save at least 500 bytes of executable.
+    // In Cargo.toml, see [profile.*.package.endoscope]
+    //
+    // TODO: Add to or create relevant cargo bug(s):
+    // e.g. https://github.com/rust-lang/cargo/issues/11680
+    //
+    #[cfg(not(clippy))]
+    if bin_size > 6 * 1024 {
+        bail!("bin_size of {bin_size} is over 6KiB. Was it built with the wrong profile?");
+    }
+
+    // Test our assumptions that the sections are in order and contiguous.
+    if sections.len() > 1 {
+        for index in 0..(sections.len() - 1) {
+            if sections[index].vm_range().end
+                != sections[index + 1].vm_range().start
+            {
+                bail!("discontiguous sections {} and {}", index, index + 1);
+            }
+        }
+    }
+
+    // We're also assuming that `vm_range`s go from 0 to bin_size.
+    // So, expect a panic if we now try to read outside those limits.
+    let mut bin = vec![0u8; bin_size];
+
+    for section in sections {
+        let file_range = section.file_range().unwrap();
+        let vm_range = section.vm_range();
+        reader.seek(SeekFrom::Start(file_range.start as u64))?;
+        if reader.read(&mut bin[vm_range.start..vm_range.end]).is_err() {
```

Comment:
```suggestion
        if reader.read_exact(&mut bin[vm_range.start..vm_range.end]).is_err() {
```
(the current code would silently allow for partial reads)